awx/ui_next/src/components/LaunchPrompt/steps/useSurveyStep.jsx

```diff
@@ -0,0 +1,119 @@
+import SurveyStep from './SurveyStep';
+import StepName from './StepName';
+
+const STEP_ID = 'survey';
+
+export default function useSurveyStep(config, resource, visitedSteps, i18n) {
+  const [stepErrors, setStepErrors] = useState({});
+
+  const { result: survey, request: fetchSurvey, isLoading, error } = useRequest(
+    useCallback(async () => {
+      if (!config.survey_enabled) {
+        return {};
+      }
+      const { data } =
+        resource.type === 'workflow_job_template'
+          ? await WorkflowJobTemplatesAPI.readSurvey(resource.id)
+          : await JobTemplatesAPI.readSurvey(resource.id);
+      return data;
+    }, [config.survey_enabled, resource])
+  );
+
+  useEffect(() => {
+    fetchSurvey();
+  }, [fetchSurvey]);
+
+  const validate = values => {
+    if (!config.survey_enabled || !survey || !survey.spec) {
+      return {};
+    }
+    const errors = {};
+    survey.spec.forEach(question => {
+      const errMessage = validateField(
+        question,
+        values[`survey_${question.variable}`],
+        i18n
+      );
+      if (errMessage) {
+        errors[`survey_${question.variable}`] = errMessage;
+      }
+    });
+    setStepErrors(errors);
+    return errors;
+  };
+
+  const hasErrors = visitedSteps[STEP_ID] && Object.keys(stepErrors).length > 0;
+
+  return {
+    step: getStep(config, survey, hasErrors, i18n),
+    initialValues: getInitialValues(config, survey),
+    validate,
+    survey,
+    isReady: !isLoading && !!survey,
+    error,
+    setTouched: setFieldsTouched => {
+      if (!survey) {
+        return;
+      }
+      const fields = {};
+      survey.spec.forEach(question => {
+        fields[`survey_${question.variable}`] = true;
+      });
+      setFieldsTouched(fields);
+    },
+  };
+}
+
+function validateField(question, value, i18n) {
+  const isTextField = ['text', 'textarea'].includes(question.type);
+  const isNumeric = ['integer', 'float'].includes(question.type);
+  if (isTextField && (value || value === 0)) {
+    if (question.min && value.length < question.min) {
+      return i18n._(t`This field must be at least ${question.min} characters`);
+    }
+    if (question.max && value.length > question.max) {
+      return i18n._(t`This field must not exceed ${question.max} characters`);
+    }
+  }
+  if (isNumeric && (value || value === 0)) {
+    if (value < question.min || value > question.max) {
+      return i18n._(
+        t`This field must be a number and have a value between ${question.min} and ${question.max}`
+      );
+    }
+  }
+  if (question.required && !value && value !== 0) {
+    return i18n._(t`This field must not be blank`);
+  }
+  return null;
+}
+
+function getStep(config, survey, hasErrors, i18n) {
+  if (!config.survey_enabled) {
+    return null;
+  }
+  return {
+    id: STEP_ID,
+    name: <StepName hasErrors={hasErrors}>{i18n._(t`Survey`)}</StepName>,
+    component: <SurveyStep survey={survey} i18n={i18n} />,
```

Comment:
(We talked about this yesterday at team sync)

I don't love that we're generating and returning components from hooks, but after pulling this down and fiddling with the code I'm starting to see why this approach is practical. 

The patternfly wizard component expects configuration objects w/ components embedded in them, so to separate the components from the hooks we would need to maintain a mapping between hook state and components somewhere else, which would add a non-trivial amount of code. It would basically be a wizard config generator with it's own config :laughing:.

This approach works. lgtm :+1: